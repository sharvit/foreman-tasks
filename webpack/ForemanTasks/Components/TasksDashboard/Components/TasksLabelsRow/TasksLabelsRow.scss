@import "~patternfly/dist/sass/patternfly/_color-variables.scss";
@import "~patternfly-react/dist/sass/_compound-label.scss";

.tasks-labels-row {
  margin: 0;
  padding: 0 10px;

  .compound-label-pf {
    margin-left: 0;
  }
}
