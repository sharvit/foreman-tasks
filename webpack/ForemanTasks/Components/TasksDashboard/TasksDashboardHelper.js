import URI from 'urijs';
import { getURIQuery } from 'foremanReact/common/helpers';

import { TASKS_DASHBOARD_AVAILABLE_TIMES_TEXT } from './TasksDashboardConstants';

export const getTimeText = time => TASKS_DASHBOARD_AVAILABLE_TIMES_TEXT[time];

export const getQueryFromUrl = () => {
  const query = {};
  const uriQuery = getURIQuery(window.location.href);

  ['state', 'result', 'mode', 'time'].forEach(f => {
    if (uriQuery[f]) query[f] = uriQuery[f];
  });

  return query;
};

export const resolveQuery = query => {
  const uri = new URI(window.location.href);
  const { search } = uri.query(true);

  const data = { search, ...query, page: 1 };
  uri.query(URI.buildQuery(data, true));
  window.Turbolinks.visit(uri.toString());
};
