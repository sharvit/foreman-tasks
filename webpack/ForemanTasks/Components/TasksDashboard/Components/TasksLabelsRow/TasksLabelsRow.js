import React from 'react';
import PropTypes from 'prop-types';
import { Row, Button, Label } from 'patternfly-react';
import { noop } from 'foremanReact/common/helpers';
import { translate as __ } from 'foremanReact/common/I18n';

import { getQueryKeyText, getQueryValueText } from '../../TasksDashboardHelper';
import { queryPropType } from '../../TasksDashboardPropTypes';
import './TasksLabelsRow.scss';

const TasksLabelsRow = ({ query, updateQuery }) => {
  const onDeleteClick = keyToDelete => {
    const { [keyToDelete]: deleted, ...queryWithoutDeleted } = query;
    updateQuery(queryWithoutDeleted);
  };

  const getLabelText = (key, value) => {
    const translatedKey = getQueryKeyText(key);
    const translatedValue = getQueryValueText(value);

    return `${translatedKey} = ${translatedValue}`;
  };

  const queryEntries = Object.entries(query);

  return (
    <Row className="tasks-labels-row">
      <span className="title">{__('Active Filters:')}</span>
      {queryEntries.map(([key, value]) => (
        <Label
          bsStyle="info"
          key={key}
          onRemoveClick={() => onDeleteClick(key)}
        >
          {getLabelText(key, value)}
        </Label>
      ))}
      {queryEntries.length > 0 && (
        <Button bsStyle="link" onClick={() => updateQuery({})}>
          {__('Clear All Filters')}
        </Button>
      )}
    </Row>
  );
};

TasksLabelsRow.propTypes = {
  query: queryPropType,
  updateQuery: PropTypes.func,
};

TasksLabelsRow.defaultProps = {
  query: {},
  updateQuery: noop,
};

export default TasksLabelsRow;
