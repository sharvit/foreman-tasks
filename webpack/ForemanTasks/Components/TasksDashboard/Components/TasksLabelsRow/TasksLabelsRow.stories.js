import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, object } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { Grid } from 'patternfly-react';

import { TASKS_DASHBOARD_AVAILABLE_TIMES } from '../../TasksDashboardConstants';
import TasksLabelsRow from './TasksLabelsRow';

storiesOf('TasksDashboard/TasksLabelsRow', module)
  .addDecorator(withKnobs)
  .addDecorator(storyFn => <Grid style={{ padding: 40 }}>{storyFn()}</Grid>)
  .add('TasksLabelsRow', () => (
    <TasksLabelsRow
      query={object('query', {
        state: 'running',
        mode: 'last',
        time: TASKS_DASHBOARD_AVAILABLE_TIMES.H24,
      })}
      deleteQueryKey={action('deleteQueryKey')}
    />
  ));
