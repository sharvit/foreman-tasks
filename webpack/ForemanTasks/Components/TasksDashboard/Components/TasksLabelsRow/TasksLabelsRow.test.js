import React from 'react';
import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';
import { shallow } from 'enzyme';

import { getQueryKeyText, getQueryValueText } from '../../TasksDashboardHelper';
import TasksLabelsRow from './TasksLabelsRow';

jest.mock('../../TasksDashboardHelper');

getQueryKeyText.mockImplementation(val => val);
getQueryValueText.mockImplementation(val => val);

const fixtures = {
  'render with minimal props': {},
  'render with props': {
    query: { some: 'query' },
    updateQuery: jest.fn(),
  },
};

describe('TasksLabelsRow', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(TasksLabelsRow, fixtures));

  describe('triggering', () => {
    it('should trigger updateQuery when label delete-click', () => {
      const updateQuery = jest.fn();
      const query = { some: 'query', someOther: 'some-query' };

      const component = shallow(
        <TasksLabelsRow query={query} updateQuery={updateQuery} />
      );
      const labels = component.find('Label');

      const firstLabel = labels.first();
      const secondLabel = labels.at(1);

      firstLabel.simulate('removeClick');
      expect(updateQuery).toHaveBeenCalledWith({ someOther: 'some-query' });

      secondLabel.simulate('removeClick');
      expect(updateQuery).toHaveBeenCalledWith({ some: 'query' });
    });

    it('should trigger updateQuery when -clear-all- click', () => {
      const updateQuery = jest.fn();
      const query = { some: 'query', someOther: 'some-query' };

      const component = shallow(
        <TasksLabelsRow query={query} updateQuery={updateQuery} />
      );
      component.find('Button').simulate('click');

      expect(updateQuery).toHaveBeenCalledWith({});
    });
  });
});
