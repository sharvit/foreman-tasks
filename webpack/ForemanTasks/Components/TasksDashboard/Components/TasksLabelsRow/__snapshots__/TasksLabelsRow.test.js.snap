// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksLabelsRow rendering render with minimal props 1`] = `
<Row
  bsClass="row"
  className="tasks-labels-row"
  componentClass="div"
>
  <span
    className="title"
  >
    Active Filters:
  </span>
</Row>
`;

exports[`TasksLabelsRow rendering render with props 1`] = `
<Row
  bsClass="row"
  className="tasks-labels-row"
  componentClass="div"
>
  <span
    className="title"
  >
    Active Filters:
  </span>
  <Label
    bsStyle="info"
    key="some"
    onRemoveClick={[Function]}
    type="default"
  >
    some = query
  </Label>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="link"
    disabled={false}
    onClick={[Function]}
  >
    Clear All Filters
  </Button>
</Row>
`;
